题目：有四个数字：1、2、3、4，能组成多少个互不相同且无重复数字的三位数？各是多少？

In [1]:
# 判断n位数是否有相同的数字
def isRep(num):
    return len(set(str(num)))==len(str(num))

In [4]:
lst = [1,2,3,4,5]
rlst = [i*100+j*10+k for i in lst for j in lst for k in lst]
for num in rlst:
    if isRep(num):
        print(num, end=" ")

123 124 125 132 134 135 142 143 145 152 153 154 213 214 215 231 234 235 241 243 245 251 253 254 312 314 315 321 324 325 341 342 345 351 352 354 412 413 415 421 423 425 431 432 435 451 452 453 512 513 514 521 523 524 531 532 534 541 542 543 

题目：企业发放的奖金根据利润提成。利润(I)低于或等于10万元时，奖金可提10%；利润高于10万元，低于20万元时，低于10万元的部分按10%提成，高于10万元的部分，可提成7.5%；20万到40万之间时，高于20万元的部分，可提成5%；40万到60万之间时高于40万元的部分，可提成3%；60万到100万之间时，高于60万元的部分，可提成1.5%，高于100万元时，超过100万元的部分按1%提成，从键盘输入当月利润I，求应发放奖金总数？

In [5]:
i = int(input('净利润:'))
arr = [1000000,600000,400000,200000,100000,0]
rat = [0.01,0.015,0.03,0.05,0.075,0.1]
r = 0
for idx in range(0,6):
    if i>arr[idx]:
        r+=(i-arr[idx])*rat[idx]
        print((i-arr[idx])*rat[idx])
        i=arr[idx]
print(r)

净利润:160000
4500.0
10000.0
14500.0


题目：一个整数，它加上100后是一个完全平方数，再加上168又是一个完全平方数，请问该数是多少？

程序分析：

假设该数为 x。

1、则：x + 100 = n2, x + 100 + 168 = m2

2、计算等式：m2 - n2 = (m + n)(m - n) = 168

3、设置： m + n = i，m - n = j，i * j =168，i 和 j 至少一个是偶数

4、可得： m = (i + j) / 2， n = (i - j) / 2，i 和 j 要么都是偶数，要么都是奇数。

5、从 3 和 4 推导可知道，i 与 j 均是大于等于 2 的偶数。

6、由于 i * j = 168， j>=2，则 1 < i < 168 / 2 + 1。

7、接下来将 i 的所有数字循环计算即可。

In [6]:
for i in range(1,85):
    if 168 % i == 0:
        j = 168 / i;
        if  i > j and (i + j) % 2 == 0 and (i - j) % 2 == 0 :
            m = (i + j) / 2
            n = (i - j) / 2
            x = n * n - 100
            print(x)

-99.0
21.0
261.0
1581.0


题目：输入某年某月某日，判断这一天是这一年的第几天？

程序分析：以3月5日为例，应该先把前两个月的加起来，然后再加上5天即本年的第几天，特殊情况，闰年且输入月份大于2时需考虑多加一天：

In [15]:
year = int(input('year:\n'))
month = int(input('month:\n'))
day = int(input('day:\n'))

m = [0,31,28,31,30,31,30,31,31,30,31,30]
months=list(np.cumsum(m))
    
#months = (0,31,59,90,120,151,181,212,243,273,304,334)
if 0 < month <= 12:
    sum = months[month - 1]
else:
    print('data error')
sum += day
leap = 0
if (year % 400 == 0) or ((year % 4 == 0) and (year % 100 != 0)):
    leap = 1
if (leap == 1) and (month > 2):
    sum += 1
print('it is the %dth day.' % sum)

year:
2020
month:
7
day:
16
it is the 198th day.


题目：斐波那契数列。

程序分析：斐波那契数列（Fibonacci sequence），又称黄金分割数列，指的是这样一个数列：0、1、1、2、3、5、8、13、21、34、……。

In [16]:
def fib(n):
    a,b = 1,1
    for i in range(n-1):
        a,b = b,a+b
    return a
 
# 输出了第10个斐波那契数列
print(fib(10))

55


In [18]:
# 使用递归
def fib(n):
    if n==1 or n==2:
        return 1
    return fib(n-1)+fib(n-2)
 
# 输出了第10个斐波那契数列
print(fib(10))

55


In [17]:
def fib(n):
    if n == 1:
        return [1]
    if n == 2:
        return [1, 1]
    fibs = [1, 1]
    for i in range(2, n):
        fibs.append(fibs[-1] + fibs[-2])
    return fibs
 
# 输出前 10 个斐波那契数列
print(fib(10))

[1, 1, 2, 3, 5, 8, 13, 21, 34, 55]


题目：输出 9*9 乘法口诀表。

程序分析：分行与列考虑，共9行9列，i控制行，j控制列。

In [23]:
for i in range(1, 10):
    print() 
    for j in range(1, i+1):
        print("{}*{}={}".format(i, j, i*j),end=" ")


1*1=1 
2*1=2 2*2=4 
3*1=3 3*2=6 3*3=9 
4*1=4 4*2=8 4*3=12 4*4=16 
5*1=5 5*2=10 5*3=15 5*4=20 5*5=25 
6*1=6 6*2=12 6*3=18 6*4=24 6*5=30 6*6=36 
7*1=7 7*2=14 7*3=21 7*4=28 7*5=35 7*6=42 7*7=49 
8*1=8 8*2=16 8*3=24 8*4=32 8*5=40 8*6=48 8*7=56 8*8=64 
9*1=9 9*2=18 9*3=27 9*4=36 9*5=45 9*6=54 9*7=63 9*8=72 9*9=81 

题目：判断101-200之间有多少个素数，并输出所有素数。

程序分析：判断素数的方法：用一个数分别去除2到sqrt(这个数)，如果能被整除，则表明此数不是素数，反之是素数。 　

题目：打印出所有的"水仙花数"，所谓"水仙花数"是指一个三位数，其各位数字立方和等于该数本身。例如：153是一个"水仙花数"，因为153=1的三次方＋5的三次方＋3的三次方。

程序分析：利用for循环控制100-999个数，每个数分解出个位，十位，百位。

In [24]:
for n in range(100,1000):
    i = n // 100
    j = n // 10 % 10
    k = n % 10
    if n == i*i*i + j*j*j + k*k*k: 
        print(n)

153
370
371
407


题目：输入一行字符，分别统计出其中英文字母、空格、数字和其它字符的个数。

程序分析：利用 while 或 for 语句,条件为输入的字符不为 '\n'。

题目：求s=a+aa+aaa+aaaa+aa...a的值，其中a是一个数字。例如2+22+222+2222+22222(此时共有5个数相加)，几个数相加由键盘控制。

程序分析：关键是计算出每一项的值。

In [36]:
Tn = 0
Sn = []
n = int(input('n = '))
a = int(input('a = '))
for count in range(n):
    Tn = Tn + a
    a = a * 10
    Sn.append(Tn)
    print(Tn)

s = 0
for i in Sn:
    s += i
print("计算和为：",s)

n = 3
a = 5
5
55
555
计算和为： 615


题目：一个数如果恰好等于它的因子之和，这个数就称为"完数"。例如6=1＋2＋3.编程找出1000以内的所有完数。

In [28]:
for j in range(2,1001):
    k = []
    n = -1
    s = j
    for i in range(1,j):
            if j % i == 0:
                n += 1
                s -= i
                k.append(i)
    
    if s == 0:
        print(j)
        for i in range(n):
            print(str(k[i]), end=" ")
        print(k[n])

6
1 2 3
28
1 2 4 7 14
496
1 2 4 8 16 31 62 124 248


题目：一球从100米高度自由落下，每次落地后反跳回原高度的一半；再落下，求它在第10次落地时，共经过多少米？第10次反弹多高？

In [1]:
tour = []
height = []
 
hei = 100.0 # 起始高度
tim = 10 # 次数
 
for i in range(1, tim + 1):
    # 从第二次开始，落地时的距离应该是反弹高度乘以2（弹到最高点再落下）
    if i == 1:
        tour.append(hei)
    else:
        tour.append(2*hei) 
    hei /= 2
    height.append(hei)

print('总高度：tour = {0}'.format(sum(tour)))
print('第10次反弹高度：height = {0}'.format(height[-1]))

总高度：tour = 299.609375
第10次反弹高度：height = 0.09765625


题目：猴子吃桃问题：猴子第一天摘下若干个桃子，当即吃了一半，还不瘾，又多吃了一个第二天早上又将剩下的桃子吃掉一半，又多吃了一个。以后每天早上都吃了前一天剩下的一半零一个。到第10天早上想再吃时，见只剩下一个桃子了。求第一天共摘了多少。

程序分析：采取逆向思维的方法，从后往前推断。

In [2]:
x2 = 1
for day in range(9,0,-1):
    x1 = (x2 + 1) * 2
    x2 = x1
print(x1)

1534


题目：两个乒乓球队进行比赛，各出三人。甲队为a,b,c三人，乙队为x,y,z三人。已抽签决定比赛名单。有人向队员打听比赛的名单。a说他不和x比，c说他不和x,z比，请编程序找出三队赛手的名单。

In [18]:
res = ["xyz","xzy","yxz","yzx","zxy","zyx"]
for i in res:
        if i[0]!="x" and i[2]!="x" and i[2]!="z":
            print("a-{},b-{},c-{}".format(i[0],i[1],i[2]))


a-z,b-x,c-y


In [16]:
import itertools
 
for i in itertools.permutations('xyz'):
    if i[0] != 'x' and i[2] != 'x' and i[2] != 'z':
        print('a vs %s, b vs %s, c vs %s' % (i[0], i[1], i[2]))

a vs z, b vs x, c vs y


In [20]:
for i in range(4):
    for j in range(2 - i + 1):
        print(' ',end="")
    for k in range(2 * i + 1):
        print('*',end="")
    print()
for i in range(3):
    for j in range(i + 1):
        print(' ',end="")
    for k in range(4 - 2 * i + 1):
        print('*',end="")
    print()

   *
  ***
 *****
*******
 *****
  ***
   *


题目：有一分数序列：2/1，3/2，5/3，8/5，13/8，21/13...求出这个数列的前20项之和。

程序分析：请抓住分子与分母的变化规律。

In [26]:
a = 2.0
b = 1.0
s = 0
for n in range(1,21):
    s += a / b
    print(str(a) + "/" + str(b), end=" ")
    t = a
    a = a + b
    b = t
print()
print(s)

2.0/1.0 3.0/2.0 5.0/3.0 8.0/5.0 13.0/8.0 21.0/13.0 34.0/21.0 55.0/34.0 89.0/55.0 144.0/89.0 233.0/144.0 377.0/233.0 610.0/377.0 987.0/610.0 1597.0/987.0 2584.0/1597.0 4181.0/2584.0 6765.0/4181.0 10946.0/6765.0 17711.0/10946.0 
32.66026079864164


题目：求1+2!+3!+...+20!的和。

程序分析：此程序只是把累加变成了累乘。

In [27]:
n = 0
s = 0
t = 1
for n in range(1,21):
    t *= n
    s += t
print('1! + 2! + 3! + ... + 20! = {}'.format(s))

1! + 2! + 3! + ... + 20! = 2561327494111820313


**(递归)** 题目：利用递归函数调用方式，将所输入的5个字符，以相反顺序打印出来。

In [28]:
def output(s,l):
    if l==0:
        return
    print(s[l-1])
    output(s,l-1)
s = input('Input a string:')
l = len(s)
output(s,l)

Input a string:hello
o
l
l
e
h
